target/linux/realtek/files-5.10/drivers/net/dsa/rtl83xx/dsa.c

```diff
@@ -911,9 +911,39 @@ static void rtl93xx_phylink_mac_link_up(struct dsa_switch *ds, int port,
 				   bool tx_pause, bool rx_pause)
 {
 	struct rtl838x_switch_priv *priv = ds->priv;
+	u32 v;
 
 	/* Restart TX/RX to port */
 	sw_w32_mask(0, 0x3, priv->r->mac_port_ctrl(port));
+
+	if (port == priv->cpu_port) {
+		if (priv->family_id == RTL9300_FAMILY_ID)
+			sw_w32(0x217, priv->r->mac_force_mode_ctrl(port));
+		if (priv->family_id == RTL9310_FAMILY_ID)
+			sw_w32(0x2a1d, priv->r->mac_force_mode_ctrl(port));
```

Comment:
Yuck, magic values.

Reviewing the [RTL930x](https://svanheule.net/realtek/longan/register/mac_force_mode_ctrl) and [RTL931x](https://svanheule.net/realtek/mango/register/mac_force_mode_ctrl) register layouts, this corresponds to the following bits:
* `MAC_FORCE_EN` (930x)
* `FORCE_LINK_EN` (930x) or  `SMI_FORCE_LINK_EN` + `SMI_FORCE_LINK` (931x)
* `DUP_SEL` (930x) or `SMI_FORCE_DUP_EN` + `SMI_DUP_SEL` (931x)
* `MAC_FORCE_FC_EN` (930x) or `SMI_FORCE_FC_EN` (931x)
* `SPD_SEL` (930x) or `SMI_FORCE_SPD_EN` + `SMI_SPD_SEL` (931x), with `(SMI_)SPD_SEL` set to 0x2 (4-bit)

If anything, at least these values appear to be consistent.

Reply:
Ok, let me double check these and add macro's for those. I (wrongfully) assumed these to be magics from the SDK

Reply:
They could still be magic values from the SDK. But since we have register field definitions, we might as well try to understand what's going on :)